### ISSUE TITLE
More fine-grained control of cache setting/getting

### ISSUE THREAD

**Issue by kentcdodds** [tannerlinsley/react-query#131]
Sorry this is a thin issue, but you said it's ok so here it is 😅

I get resources back from `/list-items` and `/list-item/{id}` and I don't want to have to re-request a list item if it's already in the cache from the `/list-items` query. Also, I can do `/list-items?bookId={bookId}` and that'll return a single list-item and I want that to resolve to a list-item I have in the cache, or set one in the cache if it's not there yet.

So basically, I want to specify a resource type more generally than just the key that is provided for me.

I'm not explaining this well, but I might be able to later...

**Comment by nikoskleidis** [comment]
I agree that this would be nice to have out of the box. 
My current solution is to prepare the cache if I know that the user is going to navigate to a specific item page.
Since I know the current item from the list, I have a "prepare" function
```js
const addItemToRQCache = async (item) =>
  prefetchQuery(["item", { id: item.id }], async () => item)
```
and before the user leaves the items listing page

```js
itemClickHandler: async (item) => {
   await addItemToRQCache(item)
   history.push(`list-items/${item.id}`)
} 
```
In the `list-items/${item.id}` page I know that there is a 
```js
useQuery(["item", { id: idFromLocation }])
```
so it will be displayed right away

**Comment by tannerlinsley** [comment]
V1 of React Query will expose a new `queryCache` object, which will be a first-class citizen with documentation, etc. It will allow you to synchronously read and write to the cache, perform batch operations (like refetching/updating/removing multiple query's caches) and will eventually allow you to customize the existing cache implementation or write your own if you really want to.  Stay tuned.

**Comment by tannerlinsley** [comment]
You can see the new api now at https://github.com/tannerlinsley/react-query/tree/next. Let me know your feedback!

**Comment by tannerlinsley** [comment]
An example of how the new queryCache can be used to achieve something like this:

```js
import { useQuery, queryCache } from 'react-query'

function Book({ bookId }) {
  // Fall back to the normal book fetcher
  useQuery(['book', bookId], (_, bookId) => {
    fetchTodoById(id)
  }, {
    // Use the already-cached book from the 'books' query as
    // the initialData if it exists
    initialData: () => {
      return queryCache.getQueryData('books')?.find(d => d.id === bookId)
    },
  })
}

```

**Comment by iamchanii** [comment]
I looked at new api. `queryCache.removeQueries` is what I want! currently `clearQueryCache` is clear all queries. indeed, in something case, should clear queries what related with a resource.

For now, I'm using `clearQueryCache` as below:

```ts
clearQueryCache();
refetch();
```

If `next` branch has released, I have to change like:

```ts
queryCache.removeQueries(resourceName); // now, remove queries what related with it.
refetch();
```

`queryCache` must be helpful with more cases. thanks you for your works, @tannerlinsley !

**Comment by tannerlinsley** [comment]
The new v1 api's should sufficiently cover these use cases. Thanks everyone for your help!

**Comment by vjuux** [comment]
@tannerlinsley @TkDodo Any update on this sweet `updateQueries` method mentioned here https://github.com/tannerlinsley/react-query/issues/135#issuecomment-589710815? I have a use case where data from WebSocket connection would need to update multiple queries, not sure how to handle this gracefully 😬

**Comment by TkDodo** [comment]
so you want to update multiple queries with the same data and match it fuzzily? I can confirm that this works:

```
        queryClient
                .getQueryCache()
                .findAll(key)
                .forEach(({ queryKey }) => {
                    queryClient.setQueryData(
                        queryKey,
                        newData
                    )
                })
```

`findAll` takes the usual fuzzy query key matching, and returns queries where you can extract the exact key from it.

**Comment by vjuux** [comment]
> so you want to update multiple queries with the same data and match it fuzzily? I can confirm that this works:
> 
> ```
>         queryClient
>                 .getQueryCache()
>                 .findAll(key)
>                 .forEach(({ queryKey }) => {
>                     queryClient.setQueryData(
>                         queryKey,
>                         newData
>                     )
>                 })
> ```
> 
> `findAll` takes the usual fuzzy query key matching, and returns queries where you can extract the exact key from it.

Thank you @TkDodo! This is what I'm currently doing, `updateQueries` seemed like a nicer alternative
